server/sonar-server/src/main/java/org/sonar/server/test/ws/TestsListAction.java

```diff
@@ -0,0 +1,221 @@
+    for (CoveredFileDoc coveredFile : coveredFiles) {
+      numberOfLinesCovered += coveredFile.coveredLines().size();
+    }
+
+    return numberOfLinesCovered;
+  }
+
+  private Map<String, ComponentDto> buildComponentsByTestFileUuid(List<TestDoc> tests) {
+    List<String> fileUuids = Lists.transform(tests, new NonNullInputFunction<TestDoc, String>() {
+      @Override
+      protected String doApply(TestDoc testDoc) {
+        return testDoc.fileUuid();
+      }
+    });
+    DbSession dbSession = dbClient.openSession(false);
+    List<ComponentDto> components;
+    try {
+      components = dbClient.componentDao().getByUuids(dbSession, fileUuids);
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+
+    return Maps.uniqueIndex(components, new Function<ComponentDto, String>() {
+      @Override
+      public String apply(@Nullable ComponentDto componentDto) {
+        return componentDto.uuid();
+      }
+    });
+  }
+
+  private SearchResult<TestDoc> searchTests(@Nullable String testUuid, @Nullable String testFileUuid, @Nullable String testFileKey, @Nullable String sourceFileUuid,
+    @Nullable Integer sourceFileLineNumber,
+    SearchOptions searchOptions) {
+    if (testUuid != null) {
+      return testIndex.searchByTestUuid(testUuid, searchOptions);
+    } else if (testFileUuid != null) {
+      return testIndex.searchByTestFileUuid(testFileUuid, searchOptions);
+    } else if (testFileKey != null) {
+      ComponentDto testFile;
+      DbSession dbSession = dbClient.openSession(false);
+      try {
+        testFile = dbClient.componentDao().getByKey(dbSession, testFileKey);
+      } finally {
+        MyBatis.closeQuietly(dbSession);
+      }
+      return testIndex.searchByTestFileUuid(testFile.uuid(), searchOptions);
+    } else {
+      return testIndex.searchBySourceFileUuidAndLineNumber(sourceFileUuid, sourceFileLineNumber, searchOptions);
+    }
+  }
+
+  private void checkPermissions(@Nullable String testUuid, @Nullable String testFileUuid, @Nullable String testFileKey, @Nullable String sourceFileUuid) {
+    if (testUuid != null) {
+      UserSession.get().checkComponentUuidPermission(UserRole.CODEVIEWER, testIndex.searchByTestUuid(testUuid).fileUuid());
+    } else if (testFileKey != null) {
+      UserSession.get().checkComponentPermission(UserRole.CODEVIEWER, testFileKey);
+    } else {
+      UserSession.get().checkComponentUuidPermission(UserRole.CODEVIEWER, firstNonNull(testFileUuid, sourceFileUuid));
+    }
+  }
+
+  private void checkArguments(@Nullable String testUuid, @Nullable String testFileUuid, @Nullable String testFileKey, @Nullable String sourceFileUuid,
+    @Nullable Integer sourceFileLineNumber) {
+    checkArgument(testUuid != null ^ testFileUuid != null ^ testFileKey != null ^ (sourceFileUuid != null && sourceFileLineNumber != null),
+      "One of the following combination must be provided: test UUID; test file UUID; test file key; source file UUID and its line number.");
```

Comment:
Does "One (and only one) of the following combination of parameters must be provided: 1) test UUID. 2) test file UUID. 3) test file key. 4) source file UUID and its line number." seems better to you ?